File: prototype/Prototype.ts
class DataSet {
    private id: number;
    data: number[];

    constructor() {
        this.data = [];
        this.id = 1;
    }

    async setup() {
        console.log("Loading data from disk...");
        await new Promise(resolve => setTimeout(resolve, 1000));
        this.data = [...this.data, 1, 2, 3];
    }

    clone() {
        const clone = Object.create(this);
        clone.data = [...this.data];
        clone.id = this.id;
        return clone;
    }
}

const testPrototype = async () => {
    // Wait 5 sec to construct an object
    const d1: DataSet = new DataSet();

    // Typescrit won't allow the waiting for construction
    // We simulate an expensive data setup process
    await d1.setup();

    // Clone the object (vid clone method)
    // No need to setup the data again
    const d2 = d1.clone();

    // Test the deep copy
    d1.data = [...d1.data, 1];
    d2.data = [...d2.data, 2];

    console.log("d1: " + d1.data.join(", "));
    console.log("d2: " + d2.data.join(", "));
}

testPrototype();

File: prototype/Prototype.cpp
#include <chrono>
#include <iostream>
#include <thread>
#include <vector>

using namespace std;

class DataSet
{
private:
    int id;

public:
    vector<int> data;
    DataSet()
    {
        // wait 5 seconds
        cout << "Loading data from disk..." << endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(5000));
    }

    // Copy constructor
    DataSet(const DataSet &other)
    {
        id = other.id;
        // vector defaultly use copy assignment operator
        data = other.data;
    }

    // Copy assignment operator
    DataSet &operator=(const DataSet &other)
    {
        id = other.id;
        // vector defaultly use copy assignment operator
        data = other.data;
        return *this;
    }

    DataSet clone()
    {
        DataSet clone = *this;
        return clone;
    }
};

void testPrototype()
{
    DataSet ds1;
    DataSet ds2 = ds1.clone();
    DataSet ds3 = ds1;

    ds1.data.push_back(1);
    ds2.data.push_back(2);
    ds3.data.push_back(3);

    cout << "ds1: " << ds1.data[0] << endl;
    cout << "ds2: " << ds2.data[0] << endl;
    cout << "ds3: " << ds3.data[0] << endl;
}

int main(int argc, char **argv)
{
    testPrototype();
    return 0;
}